File: string_center/lib/tab.dart
import 'dart:convert';


class Tabb{
  String _info;
  List<Measure> _measures;

  Tabb(String info){
    _info = info;
    _measures = [new Measure("first measure", 6, 4)];
  }

  Tabb.fromJson(encodedJson){
    Map json = JSON.decode(encodedJson);
    _info = json['info'];
    num measureCount = json['measureCount'];
    _measures = [];
    for(int i = 0; i < measureCount; i++) {
      _measures.add(new Measure.fromJson(json['measures'][i]));
    }
  }

  Map toJson(){
    Map map = new Map();
    map['info'] = _info;
    map['measureCount'] = _measures.length;
    map['measures'] = _measures;
    return map;
  }

  void addMeasure(String info, int stringCount, int noteCount){
    _measures.add(new Measure(info, stringCount, noteCount));
  }

  void printTabb(){
    for(Measure measure in _measures){
      for(InstString string in measure.strings){
        for(String note in string.notes){
          print(note);
        }
      }
    }
  }

  get info => _info;
  get measureCount => _measures.length;
}


class Measure{
  String _info;
  List<InstString> _strings;

  Measure(String info, int stringCount, int noteCount){
    _info = info;
    _strings = [];
    for(int i = 0; i < stringCount; i++){
      _strings.add(new InstString(noteCount));
    }
  }

  Measure.fromJson(json){
    _info = json['info'];
    num stringCount = json['stringCount'];
    _strings = [];
    for(int i = 0; i < stringCount; i++){
      _strings.add(new InstString.fromJson(json['strings'][i]));
    }
  }

  Map toJson(){
    Map map = new Map();
    map["info"] = _info;
    map['stringCount'] = _strings.length;
    map['strings'] = _strings;
    return map;
  }

  void addString(int noteCount){
    _strings.add(new InstString(noteCount));
  }

  get info => _info;
  get stringCount => _strings.length;
  get strings => _strings;
}


class InstString  {
  List<String> _notes;

  InstString(int noteCount){
    _notes = [];
    for(int i = 0; i < noteCount; i++){
      _notes.add('-');
    }
  }

  InstString.fromJson(json){
    _notes = json['notes'];
  }

  void addNote(){
    _notes.add('-');
  }

  Map toJson(){
    Map map = new Map();
    map["notes"] = _notes;
    return map;
  }

  get noteCount => _notes.length;
  get notes => _notes;
}

File: StringCenterNode/public/javascripts/tab/tab.js
function Tab(tabInfo){
  this.info = tabInfo.name;
  this.measureCount = 1;
  this.measures = [];
  if(this.validTabInfo(tabInfo)){
      this.initTab(tabInfo.name, tabInfo.tuning.length, tabInfo.tuning, tabInfo.noteCount);
  }
}
function Measure(info, stringCount, tuning){
  this.info = info;
  this.stringCount = stringCount;
  this.strings = [];
  this.tuning = tuning;
}

function InstString(tuning, noteCount){
  this.tuning = tuning;
  this.noteCount = noteCount;
  this.notes = [];
}

Tab.prototype.validTabInfo = function(i){
    return (i.name && i.name.length < 100 && i.tuning && i.tuning.length >= 4 && i.noteCount && parseInt(i.noteCount) < 32);
}

Tab.prototype.initTab = function(info, stringCount, tuning, noteCount){
  this.addMeasure(undefined, info, stringCount, tuning, noteCount);
}

Tab.prototype.addMeasure = function(measure, info, stringCount, tuning, noteCount){
    if(checkMeasure(measure)){
        this.measures.push(measure);
    } else {
        this.measures.push(new Measure(info, stringCount, tuning));
        this.measures[this.measures.length - 1].initMeasure(noteCount);
        this.measureCount = this.measures.length;
    }
}

Tab.prototype.addNote = function(measureNum, note, noteNum, stringNum){
  this.measures[measureNum].strings[stringNum].notes[noteNum] = note;
}

Tab.prototype.displayTab = function(){
  for(var i = 0; i < this.getMeasureCount(); i++){
    console.log("Measure " + i + ": " + this.measures[i].info);
    for(var j = 0; j < this.getStringCount(); j++){
      var instStringToBePrinted = this.measures[i].strings[j].tuning;
      var instStringToBePrinted = this.measures[i].strings[j].tuning + " ";
      for(var k = 0; k < this.getNoteCountPerString(); k++){
        instStringToBePrinted += this.measures[i].strings[j].notes[k];
      }
      console.log(instStringToBePrinted);
    }
    console.log("");
  }
}

Tab.prototype.getStringCount = function(){
  return this.measures[0].stringCount;
}

Tab.prototype.getMeasureCount = function(){
  return this.measureCount;
}

Tab.prototype.getNoteCountPerString = function(){
  return this.measures[0].strings[0].noteCount;
}

Tab.prototype.getTuning = function(){
  return this.measures[0].tuning;
}

Measure.prototype.initMeasure = function(noteCount){
  for(var i = 0; i < this.stringCount; i++){
    this.strings.push(new InstString(this.tuning[(this.tuning.length - 1) - i], noteCount));
    this.strings[i].initString(noteCount);
  }
}

function checkMeasure(measure){
    return (measure && measure.info && measure.info.length < 100 &&
            measure.stringCount && measure.stringCount > 0 && measure.stringCount < 9 &&
            measure.strings.length > 0 && measure.strings.length < 9 &&
            measure.tuning && measure.tuning.length == measure.strings.length);
}

InstString.prototype.initString = function(noteCount){
  for(var i = 0; i < noteCount; i++){
    this.notes.push("-");
  }
}
